### ISSUE TITLE
In Azure ARM REST API, how do a get the state (i.e. started, stopped etc) for running virtual machines?

### ISSUE THREAD

**Issue by dukedougal** [Azure/azure-xplat-cli#2565]
I am using ARM in Azure.

Azure provides a REST API to get a list of virtual machines.

https://msdn.microsoft.com/en-us/library/azure/mt163572.aspx

Strangely the response data does not seem to provide the state of those machines i.e stopped, running etc. Am I looking in the wrong place?

I need to be able to get a list of all virtual machines and carry out actions on them depending on whether or not they are running or stopped etc.

**Comment by amarzavery** [comment]
@dukedougal - I assume that you would have read my response about changing modes in the previous issue had filed https://github.com/Azure/azure-xplat-cli/issues/2566 

So, if you execute the vm list in arm mode then the output has PowerState property 
```
D:\sdk\azure-xplat-cli>azure vm list
info:    Executing command vm list
+ Getting virtual machines
data:    ResourceGroupName      Name             ProvisioningState  PowerState      Location       Size
data:    ---------------------  ---------------  -----------------  --------------  -------------  -----------
data:    TST232                  test12             Succeeded          VM running      westus         Standard_A1
data:    TST343                  test34             Failed             VM running      westus         Standard_A1
info:    vm list command OK
```

**Comment by dukedougal** [comment]
Hello @amarzavery - I am trying to program only using the ARM REST API.

**Comment by amarzavery** [comment]
@dukedougal - Are you using the CLI or the node sdk?

**Comment by amarzavery** [comment]
@dukedougal - This the CLI project not the node sdk project. I have shown you the output of `azure vm list` command that has the PowerState property in its output.

**Comment by dukedougal** [comment]
Hi @amarzavery - I'm trying to use only the REST API and it appears as far as I can tell that the cli does an initial request for the list of VMs then sends an additional request for every VM to get the required state data.

**Comment by amarzavery** [comment]
@dukedougal - I understand what you are doing. The script that you are writing is from the (azure-arm-compute) node sdk which is in this project https://github.com/Azure/azure-sdk-for-node and you are filing an issue in xplat cli which consumes the node sdk. These are two different things. As far as ARM API for VM is not offering parity wrt V1 API, the service team would be able to comment on it.

@huangpf , @sauryadas , @singhkay - Please respond to this customer query.

**Comment by singhkay** [comment]
@dukedougal You are correct, with the ARM API you can't get the VM state in the List ALL VMs API. You need to call the instance view API on each VM to get this. 

`https://management.azure.com/subscriptions/{subscription-id}/resourceGroups/{resource-group-name}/providers/Microsoft.Compute/virtualMachines/{vm-name}/InstanceView?api-version={api-version}`

Adding @mahthi for this feedback

**Comment by dukedougal** [comment]
On EC2 I can ask for a description of my instances and I get everything I need to know, but on Azure I need to send an individual request for every instance just to find out if they are running or stopped.  Seems like a flawed design to me. What do you guys think?

 http://docs.aws.amazon.com/AWSEC2/latest/CommandLineReference/ApiReference-cmd-DescribeInstances.html

**Comment by jdc0589** [comment]
@dukedougal It's definitely a problem. It makes application code a ton more complicated. Instead of one API call, not only do you have to do 1 + # of machines, but the invididual VM lookups have to be run run async with a worker pool if you want any kind of reasonable performance.

It's a pain.

**Comment by gkapagunta** [comment]
> @dukedougal You are correct, with the ARM API you can't get the VM state in the List ALL VMs API. You need to call the instance view API on each VM to get this.

> https://management.azure.com/subscriptions/{subscription-id}/resourceGroups/{resource-group-name}/providers/Microsoft.Compute/virtualMachines/{vm-name}/InstanceView?api-version={api-version}

>Adding @mahthi for this feedback


Hi @singhkays can you point me to the documentation which specifies how we can get the Instance view for a VM via REST API? I could not find it anywhere, although there is documentation to get Instance view for a VM in a scale set, but that is clearly not the scenario.

**Comment by danmo** [comment]
This is a pretty important and easy to implement feature one would think.

**Comment by huangpf** [comment]
This is an ask about service API, not the CLI commands. For service issue, please use your subscription's support.

**Comment by gokhanoner** [comment]
Also, in order to get public IP, Private IP, VM list, VM status, you need to do a separate REST calls. (For V run status you need to do separate calls for each VM !!!) Is it really that hard to provide all these info in a single REST call?

**Comment by grumblegirl** [comment]
This document instruct you to pass parameter "statusOnly=true" to get VM stauts in a single call: https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/listall#code-try-0. 

And the response shows "message": "Query param statusOnly=true is only supported when subscription is registered with Microsoft.Compute/BulkGETVMRuntimeState feature."

And i ran "az feature list --namespace Microsoft.Compute --output table" on CLI. Confirmed that "Microsoft.Compute/BulkGETVMRuntimeState" is not a feature for Microsoft.Compute.

So, the official document is wrong. The API response is wrong.

And yes, when you are working with large environment, getting all the VM names first over API to send another API request to get VM running state is not a small deal.

Most importantly, I do not see an API where you can pass all VM names in a single request to get the running state. If you have 2000 VMs, you have to call 2000 REST API to VM to get the running state.

This is sub-optimal from the quality of documentation, incorrect API response showing deprecated feature as required, and no API replacing the depreciated feature, when comparing with AWS and GCP. It takes little to no effort to get this info over API from other 2 clouds.
